### DISCUSSION TITLE
Compiled bindings in .NET Maui 9, pass the getter as a parameter?

### DISCUSSION THREAD

**Discussion by JohnHDev** [dotnet/maui#26837] [Q&A]
Hi, I am investigating the use of compiled bindings in a small fluent framework Ive created. Given the following code that uses a compiled binding:

            var label = new Label();
            label.SetBinding(Label.TextProperty, static (IViewModel m) => m.Date);

I want to pass a static function to my own fluent function like this:

            Content = new Label()
                .TextBinding(static (IViewModel m) => m.Date)
                .BackgroundColor(Colors.Red);

The code for the TextBinding method looks like this:

        public static Label TextBinding<TSource, TProperty>(this Label self, Func<TSource, TProperty> getter)
        {
            self.SetBinding(Label.TextProperty, getter);
            return self;
        }

But I get the following error:

            error BSG0002: The getter must be a lambda expression

The error is quite clear, so was thinking I might need to look at the codegen to see how that is implemented. Before I do that I wanted to check if Ive missed anything.

It might be Ive just had too much turkey at this point to see the wood for the trees. 🌲

(I have simplified the code above massively to get to a small example of what Im trying to achieve)

**Comment by jkurdek** [comment]
@JohnHDev, thanks for the write up! I'm afraid passing the getter as a parameter is not supported now, the tooling expects an inline lambda expression.

If you look at the code generation, you’ll see that it uses interceptors to write code for self.SetBinding(Label.TextProperty, ...) at compile time. This process relies on the lambda’s source code to generate the correct binding code. By requiring that the lambda expression be inline, the generator can always retrieve and analyze the Roslyn expression tree. In order to support passing the getter as a parameter, we’d need to figure out how to capture the lambda’s expression tree when it’s defined elsewhere.

**Reply by JohnHDev** [reply]
Thanks I'll take a look at the codegen, can you point me in the right direction?

**Reply by jkurdek** [reply]
Sure, the logic resides here https://github.com/dotnet/maui/blob/07cc9730fcb320d3343ad0fbc9e13e65343fb654/src/Controls/src/BindingSourceGen/BindingSourceGenerator.cs#L134-L156

**Reply by JohnHDev** [reply]
@jkurdek thanks for that, I see under the hood its using TypedBinding, I have implemented that now (at least for getter, setter I'll do shortly).

So syntax wise, I now have:

        Content = new Label()
            .TextBinding(static (IViewModel m) => m.Date, (DateTime d) => d.ToString("dd MMM yyyy"))
            .BackgroundColor(Colors.Red);

The DateTime func maps onto a generic from/to property converter.

Question though, are there any advantages to using the static keyword? It isn't required but Im not aware of any speed advantages.

**Reply by jkurdek** [reply]
Yes, under the good it is TypedBindings. The BindingSourceGenerator generates the correct getter, setter and handlers array. Here is a more complex example: https://github.com/dotnet/maui/blob/46096ffcf6772f913fb14b99749a19d2e0a419d7/src/Controls/tests/BindingSourceGen.UnitTests/IntegrationTests.cs#L893

The static keyword is there to enforce that the lambda doesnt use any local variables on its rights side e.g. `var x = ..., (IViewModel m) => x` is not allowed

**Comment by AmrAlSayed0** [comment]
You can modify the `TextBinding` to this:
```csharp
    public static Label TextBinding<TSource, TProperty>(this Label self, Expression<Func<TSource, TProperty>> getter)
    {
        self.SetBinding(Label.TextProperty, getter);
        return self;
    }
```
and it should work ... I think

**Reply by JohnHDev** [reply]
> You can modify the `TextBinding` to this:
> 
> ```cs
>     public static Label TextBinding<TSource, TProperty>(this Label self, Expression<Func<TSource, TProperty>> getter)
>     {
>         self.SetBinding(Label.TextProperty, getter);
>         return self;
>     }
> ```
> 
> and it should work ... I think

ah no unfortunately not, I looked at that too.

**Comment by stephenquan** [comment]
Am I missing something here? [BindingBase.Create<TSource, TProperty>](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.bindingbase.create?view=net-maui-9.0) already supports static lambda expression getter function, for example:

```c#
label.SetBinding(Label.TextProperty, BindingBase.Create<ViewModel, DateTime>(static m => m.Date, source: viewModel));
```

**Reply by JohnHDev** [reply]
True and I used TypedBinding specifically for that, the key was being able to pass in the getter function as a parameter to a containing method. I did that using an Expression with a compile and invoke
